modules/tinymce/src/core/main/ts/lookup/UserLookup.ts

```diff
@@ -0,0 +1,214 @@
+ * });
+ */
+
+type UserId = string;
+
+export interface User {
+  id: UserId;
+  name?: string;
+  avatar?: string;
+  description?: string;
+  custom?: Record<string, any>;
+}
+
+export interface ValidatedUser {
+  id: UserId;
+  name: Optional<string>;
+  avatar: Optional<string>;
+  description: Optional<string>;
+  custom: Optional<Record<string, any>>;
+}
+
+export interface UserLookup {
+  /**
+   * Retrieves the current user ID from the editor.
+   *
+   * @method getCurrentUserId
+   * @return {string} The current user ID.
+   */
+  getCurrentUserId: () => UserId;
+
+  /**
+   * Fetches user information using a provided array of userIds.
+   *
+   * @method fetchUsers
+   * @param {string[]} userIds - A list of user IDs to fetch information for.
+   * @throws {Error} Throws an error if any of the user IDs are invalid or if the fetch fails.
+   * @return {Promise<User>[]} A promise that resolves to an array of users and information about them.
+   */
+  fetchUsers: (userIds: UserId[]) => Promise<User>[];
+}
+
+const userSchema = StructureSchema.objOf([
+  FieldSchema.required('id'),
+  FieldSchema.optionString('name'),
+  FieldSchema.optionString('avatar'),
+  FieldSchema.optionString('description'),
+  FieldSchema.option('custom')
+]);
+
+const objectCat = <T extends object>(
+  obj: { [K in keyof T]: Optional<T[K]> }
+): Partial<T> => {
+  const result = {} as Partial<T>;
+
+  Obj.each(obj, (value, key) => {
+    value.each((v) => {
+      result[key as keyof T] = v;
+    });
+  });
+
+  return result;
+};
+
+const validateResponse = (items: unknown): User[] => {
+  if (!Array.isArray(items)) {
+    throw new Error('fetch_users must return an array');
+  }
+
+  const results = Arr.map(items, (item) =>
+    StructureSchema.asRaw<ValidatedUser>('Invalid user object', userSchema, item)
+  );
+
+  const { errors, values } = Results.partition(results);
+
+  if (errors.length > 0) {
+    const formattedErrors = Arr.map(errors, (error, idx) =>
+      `User at index ${idx}: ${StructureSchema.formatError(error)}`
+    );
+
+    // eslint-disable-next-line no-console
+    console.warn('User validation errors:\n' + formattedErrors.join('\n'));
+  }
+
+  return Arr.map(values, (user) => {
+    const { id, ...rest } = user;
+
+    return {
+      id,
+      ...objectCat(rest),
+    };
+  });
+};
+
+const UserLookup = (editor: Editor): UserLookup => {
+  const userCache = new Map<UserId, Promise<User>>();
+  const pendingResolvers = new Map<UserId, {
+    resolve: (user: User) => void;
+    reject: (error: unknown) => void;
+  }>();
+
+  const lookup = (userId: UserId) =>
+    Optional.from(userCache.get(userId));
+
+  const store = (user: Promise<User>, userId: UserId) => {
+    userCache.set(userId, user);
+  };
+
+  const finallyReject = (userId: UserId, error: Error) => {
+    return Optional.from(pendingResolvers.get(userId)).each(({ reject }) => {
+      reject(error);
+      pendingResolvers.delete(userId);
+    });
+  };
```

Comment:
```suggestion
  const finallyReject = (userId: UserId, error: Error) =>
    Optional.from(pendingResolvers.get(userId)).each(({ reject }) => {
      reject(error);
      pendingResolvers.delete(userId);
    });
```
This also returns void anyway.